### PR TITLE
ARROW-9486: [C++][Dataset] Support implicit cast of InExpression::set to dict

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

File: cpp/src/arrow/dataset/filter.cc [hunk 1]
```diff
@@ -931,6 +931,22 @@ Result<std::shared_ptr<DataType>> FieldExpression::Validate(const Schema& schema
   return null();
 }
 
+Result<Datum> CastOrDictionaryEncode(const Datum& arr,
```
Comment (pitrou) [inline cpp/src/arrow/dataset/filter.cc:934]: This doesn't seem to be static nor in the anonymous namespace?
(ditto for other private functions in this file?)

**Review by pitrou — COMMENTED** [review]
Why not, but I can't rely evaluate the consequences.

**Review by wesm — APPROVED** [review]
+1

**Review by wesm — APPROVED** [review]
+1

File: cpp/src/arrow/dataset/filter.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/filter.cc:934]: https://issues.apache.org/jira/browse/ARROW-9492

**Pull request closed without merging**